gufe/chemicalsystem.py

```diff
@@ -126,6 +126,54 @@ def total_charge(self):
                 total_charge += fc
         return total_charge
 
+    def contains(self, item: Component | type[Component]) -> bool:
+        """Check if a Component or Component type is in this ChemicalSystem.
+
+        Parameters
+        ----------
+        item : Component or type of Component
+            The Component instance or class to check for.
+
+        Returns
+        -------
+        bool
+            Returns True if the item is found in the ChemicalSystem, False otherwise.
+        """
+        # check the input types
+        if not (isinstance(item, Component) or (isinstance(item, type) and issubclass(item, Component))):
+            raise TypeError("`item` must be an instance or subclass of `Component`")
+
+        for comp in self._components.values():
+            if isinstance(item, Component):
+                if comp == item:
+                    return True
+            elif issubclass(item, Component):
+                if isinstance(comp, item):
+                    return True
+
+        return False
+
+    def get_components_of_type(self, item: type[Component]) -> list[Component]:
+        """Get all Components of a specific type in this ChemicalSystem
+
+        Parameters
+        ----------
+        item : type of Component
+            The class of Component to search for.
+
+        Returns
+        -------
+        list of Component
+            A list of all Components in the ChemicalSystem that are instances of the specified type.
+            If no Components of the specified type are found, an empty list is returned.
+        """
+        if not (isinstance(item, type) and issubclass(item, Component)):
+            raise TypeError("`item` must be a subclass of `Component`")
+
+        matches = [comp for comp in self._components.values() if isinstance(comp, item)]
```

Comment:
Non-blocking thought:
Am I correct in thinking that all items will be unique here? I'm considering @ijpulidos's preference for sets (which tends to be my preference as well), but I don't think it makes much of a difference here. Plus, returning a list might make it clearer that we're *not* pulling just the unique items from a list that contains duplicates. 

Third idea just to make things harder: maybe a set would be appropriate here, since it suggests immutability?